h/api/test/auth_test.py

```diff
@@ -102,64 +94,92 @@ def test_userid_from_bearer_token_calls_decode(jwt):
+@userid_from_jwt_fixtures
+def test_userid_from_jwt_returns_None_if_no_sub(jwt):
     jwt.decode.return_value = {}  # No 'sub' key.
 
-    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+    result = auth.userid_from_jwt(u'abc123', mock_request())
 
     assert result is None
 
 
-@user_id_from_bearer_token_fixtures
-def test_userid_from_bearer_token_returns_None_if_decoding_fails(jwt):
+@userid_from_jwt_fixtures
+def test_userid_from_jwt_returns_None_if_decoding_fails(jwt):
     jwt.decode.side_effect = InvalidTokenError
 
-    result = auth.userid_from_bearer_token(mock_request(token='Bearer abc123'))
+    result = auth.userid_from_jwt(u'abc123', mock_request())
 
     assert result is None
 
 
-def test_generate_bearer_token_userid_from_bearer_token_successful():
-    """Test generate_bearer_token() and userid_from_bearer_token() together.
+def test_generate_jwt_userid_from_jwt_successful():
+    """Test generate_jwt() and userid_from_jwt() together.
 
-    Test that userid_from_bearer_token() successfully decodes tokens
-    generated by generate_bearer_token().
+    Test that userid_from_jwt() successfully decodes tokens
+    generated by generate_jwt().
 
     """
-    token = auth.generate_bearer_token(mock_request(), 3600)
-    userid = auth.userid_from_bearer_token(
-        request=mock_request(token='Bearer ' + token))
+    token = auth.generate_jwt(mock_request(), 3600)
+    userid = auth.userid_from_jwt(token, mock_request())
 
     assert userid == 'acct:testuser@hypothes.is'
 
 
-def test_generate_bearer_token_userid_from_bearer_token_bad_token():
-    """Test generate_bearer_token() and userid_from_bearer_token() together.
+def test_generate_jwt_userid_from_jwt_bad_token():
+    """Test generate_jwt() and userid_from_jwt() together.
 
-    Test that userid_from_bearer_token() correctly fails to decode a token
-    generated by generate_bearer_token() using the wrong secret.
+    Test that userid_from_jwt() correctly fails to decode a token
+    generated by generate_jwt() using the wrong secret.
 
     """
     request = mock_request()
     request.registry.settings['h.client_secret'] = 'wrong'
-    token = auth.generate_bearer_token(request, 3600)
+    token = auth.generate_jwt(request, 3600)
 
-    userid = auth.userid_from_bearer_token(
-        request=mock_request(token='Bearer ' + token))
+    userid = auth.userid_from_jwt(token, mock_request())
 
     assert userid is None
 
 
+userid_from_api_token_fixtures = pytest.mark.usefixtures('models')
+
+
+@userid_from_api_token_fixtures
+def test_userid_from_api_token_returns_None_when_token_doesnt_start_with_6879(
+        models):
+    assert auth.userid_from_api_token(u'abc123') is None
+    assert not models.Token.get_by_value.called
+
+
+@userid_from_api_token_fixtures
+def test_userid_from_api_token_calls_get_by_value(models):
+    auth.userid_from_api_token(u'6879-abc123')
+
+    models.Token.get_by_value.assert_called_once_with(u'6879-abc123')
```

Comment:
I'm with Sean here, and it might be useful if I try and explain a little of the why. You're certainly right that tests should test the interface of a unit, and not its implementation. Why? Because it makes the tests fragile -- meaning that you could refactor the unit and break the tests, even if the code actually works. That's a surefire way to make refactoring more painful (and thus less likely to happen).

But don't forget that many (but not all) units have *two* interfaces.

1. The interface the unit exposes to others -- its API contract. For a function that's basically its call signature and perhaps some aspects of the business logic it implements.
2. The dependencies your "unit under test" has on other units. Many units collaborate in order to achieve their ends -- if you don't control this second interface when testing, you're writing integrated tests, not unit test.

So, you want to verify *both* that the unit is upholding its API contract, *and* that it is correctly collaborating with other components.

This naturally leads to two kinds of unit tests: *contract* and *collaboration* tests. I'd suggest you always endeavour to keep them separated into distinct tests. This one is a collaboration test, because `userid_from_api_token` collaborates with `Token.get_by_value`, so we want to ensure that it is calling that function correctly.

I hope that goes some way to understanding the rationale. You can read more about contract/collaboration tests [here](http://bartling.blogspot.de/2012/01/understanding-power-of-isolated-object.html) and [here](http://blog.thecodewhisperer.com/permalink/stub-your-data-access-layer-it-wont-hurt). The original idea comes from [this talk](https://vimeo.com/80533536).